scio-core/src/main/scala/com/spotify/scio/io/BinaryIO.scala

```diff
@@ -206,4 +232,114 @@ object BinaryIO {
+    emptyMatchTreatment: EmptyMatchTreatment = ReadParam.DefaultEmptyMatchTreatment,
+    suffix: String = ReadParam.DefaultSuffix
+  )
+
+  /** Trait for reading binary file formats */
+  trait BinaryFileReader {
+
+    /**
+     * State constructed during `start` and maintained through reading, potentially updated during
+     * each record read
+     */
+    type State
+
+    /** Read any header and construct the initial read state. */
+    def start(is: InputStream): State
+
+    /**
+     * @return
+     *   The updated read state, record bytes. Return null for record byte array when no record is
+     *   read.
+     */
+    def readRecord(state: State, is: InputStream): (State, Array[Byte])
+
+    /** Read any footer and perform any required validation after the last record is read. */
+    def end(state: State, is: InputStream): Unit = ()
+  }
+
+  final private[scio] class BinarySingleFileSource(
+    binaryFileReader: BinaryFileReader,
+    metadata: Metadata,
+    start: Long,
+    end: Long
+  ) extends FileBasedSource[Array[Byte]](metadata, Long.MaxValue, start, end) {
+    override def isSplittable: Boolean = false
+
+    override def createForSubrangeOfFile(
+      fileMetadata: Metadata,
+      start: Long,
+      end: Long
+    ): FileBasedSource[Array[Byte]] =
+      throw new NotImplementedError()
+
+    override def createSingleFileReader(
+      options: PipelineOptions
+    ): FileBasedSource.FileBasedReader[Array[Byte]] =
+      new FileBasedSource.FileBasedReader[Array[Byte]](this) {
+        var is: CountingInputStream = _
+        var state: binaryFileReader.State = _
+        var startOfRecord: Long = _
+        var current: Option[Array[Byte]] = None
+
+        // exception matches method contract
+        override def getCurrentOffset: Long = current.map(_ => startOfRecord).get
+
+        // exception matches method contract
+        override def getCurrent: Array[Byte] = current.get
+
+        override def startReading(channel: ReadableByteChannel): Unit = {
+          is = new CountingInputStream(Channels.newInputStream(channel))
+          val newState = binaryFileReader.start(is)
+          state = newState
+        }
+
+        override def readNextRecord(): Boolean = {
+          startOfRecord = is.getCount + 1
+          val (newState, record) = binaryFileReader.readRecord(state, is)
+          state = newState
+          current = Option(record)
+          current match {
+            case Some(_) => true
+            case None =>
+              binaryFileReader.end(state, is)
+              false
+          }
+        }
+      }
+  }
+
+  final private[scio] class BinarySource(
+    path: String,
+    emptyMatchTreatment: EmptyMatchTreatment,
+    binaryFileReader: BinaryFileReader
+  ) extends FileBasedSource[Array[Byte]](
+        StaticValueProvider.of(path),
+        emptyMatchTreatment,
+        Long.MaxValue
+      ) {
+    override def isSplittable: Boolean = false
```

Comment:
wouldn't this be `true` because this source implements file subranges?

Reply:
No, the subrange API is called for the entire file, then `isSplittable` is checked to determine if further splits should occur. 

The actual format we write with `saveAsBinaryFile` is not splittable since we have no record count, no index, and no blocks of records.

Reply:
ah okay. It's funny that we don't need to implement `createSingleFileReader` then -- I would assume that the runner would invoke `createSingleFileReader` if `iSplittable == false` and `createForSubrangeOfFile` otherwise

Reply:
It's also because in javaland there's a single class with two constructors; one for the path-created variant and one for the file-variant. Here it's split into `BinarySource` and `BinarySingleFileSource`